Cross-validation and pruning

parismita opened the pull request.

Commit:
Add files via upload

Commit:
Update pruning.R

Commit:
Update pruning.R

Commit:
Add files via upload

Commit:
Merge pull request #1 from parismita/master

updating

Commit:
Update growtree.R

parismita committed:
fix bugs in weakest link

parismita committed:
fix bugs in initial pruning and create sequencial pruning function

parismita committed:
update pruning

parismita committed:
add tail to import from in namespace

parismita committed:
update pruning

parismita committed:
update pruning

parismita committed:
updating branch

Merge branch 'master' of https://github.com/parismita/mmit into prunning

Conflicts:
	.gitignore
	Rpackage/R/growtree.R

aldro61 commented on Rpackage/R/initial.pruning.R:
```diff
@@ -0,0 +1,44 @@
+    node <- as.partynode(t)
+    
+    ### left right child ids and partynodes
+    left_kid_id <- sapply(kids_node(node), id_node)[1]
```
Do `sapply(kids_node(node), id_node)` only once and use index

aldro61 replied:
Also, no need to get the ids and then cut the tree at the kids. kids_node already returns the kid nodes.

aldro61 commented on Rpackage/R/initial.pruning.R:
```diff
@@ -0,0 +1,44 @@
+    node_id <- tail(inner_id, n = n)[1]
+    ### partynode object
+    t <- nodeapply(tree, ids = node_id)[[1]]
+    node <- as.partynode(t)
```
not required

aldro61 commented on Rpackage/R/pruning.R:
```diff
@@ -0,0 +1,44 @@
+pruning <- structure(function(tree){
+  ### T1 is the tree after initital pruning of Tmax
+  T1 <- init_pruning(tree)
+  T1 <- nodeapply(T1, ids = 1)[[1]]
```
Don't think that this is requied

aldro61 commented on Rpackage/R/weakestLink.R:
```diff
@@ -0,0 +1,88 @@
+    ### get node id and cost of the current node
+    ter_id <- nodeids(tree, terminal = TRUE)
+    all_id <- nodeids(tree, terminal = FALSE)
+    node_cost <- as.numeric(nodeapply(tree, ids = all_id[1], info_node)[[1]][2])
```
Invert lines and replace by var

aldro61 commented on Rpackage/R/weakestLink.R:
```diff
@@ -0,0 +1,88 @@
+      ### current alpha value
+      current_gt <- (node_cost - C_Tt)/(length(ter_id) - 1)
+      
+      ### left right child ids and partynodes
```
See comment above about kids

parismita committed:
branch update

parismita committed:
gitignore

Merge branch 'master' of https://github.com/parismita/mmit into prunning

Conflicts:
	.gitignore

Commit:
Update growtree.R

parismita committed:
weakest link and initial pruning code update

Commit:
Merge pull request #2 from aldro61/master

updating branch and removing conflicts

Commit:
Update .gitignore

parismita committed:
name change solln to wlink

parismita committed:
create fit_and_score funcion for cross validation
create metric for score calc.

parismita committed:
bug fixing in fit and score

parismita committed:
indexing bug in metric.R fixed

parismita committed:
implementing linear search of alphas for cv_score

parismita committed:
fixed return value datatype as list in fit and score

aldro61 commented on Rpackage/R/pruning.R:
```diff
@@ -0,0 +1,49 @@
+    if(length(order(wlink[, 2]))>1){
+      wlink <- wlink[order(wlink[, 2]),]
+    }
+    print(wlink)
```
Don't forget to remove print statement when done.

parismita replied:
ok

aldro61 commented on Rpackage/R/pruning.R:
```diff
@@ -0,0 +1,49 @@
+    
+    ### prune the weakest links from tail of tree
+    ### sort solution
+    if(length(order(wlink[, 2]))>1){
```
You call order twice, so sorting twice, which is inefficient.

aldro61 replied:
Also, it would be best to add a comment stating why you are reordering the weakest links for future maintenance.

parismita replied:
ok

aldro61 commented on Rpackage/R/pruning.R:
```diff
@@ -0,0 +1,49 @@
+    }
+    print(wlink)
+    for(n in length(wlink[, 2]):1){
+      ### ignore non int value less than 1
```
How is it possible to get non-int values here and especially less than 1?

parismita replied:
it should not be possible but i was getting float values for root between 0 to 1

aldro61 commented on Rpackage/R/pruning.R:
```diff
@@ -0,0 +1,49 @@
+    tree1 <- nodeapply(tree, ids = 1)[[1]]
+    
+    ### if terminal root return current tree
+    if(is.terminal(tree1)){
```
Do you really need to create a variable for this? You only use it once.

parismita replied:
ok

aldro61 commented on Rpackage/R/pruning.R:
```diff
@@ -0,0 +1,49 @@
+  ### party object to partynode (tree1 and tree are equal)
+  Tree1 <- nodeapply(T1, ids = 1)[[1]]
+  
+  if(is.terminal(Tree1)){
```
Same comment as above on variable creation.

aldro61 commented on Rpackage/R/weakestLink.R:
```diff
@@ -0,0 +1,91 @@
+      ### left right child ids and partynodes
+      kids <- kids_node(tree)
+      kid_ids <- sapply(kids, id_node)
+      left_kid_id <- kid_ids[1]
```
These variables are not used later. Forgot to delete?

parismita replied:
yep

aldro61 commented on Rpackage/R/weakestLink.R:
```diff
@@ -0,0 +1,91 @@
+  
+      ### weakest link from left tree part
+      wlink <- find_weakest_link(kids[[1]])
+      left_min_gt <- tail(wlink)[1]
```
It's a bit obscur to me why your woul duse tail here. What does it return (data wise)?

aldro61 commented:
@parismita I did a quick code review. Did not dive into the details. Will continue tomorrow. Please add unit tests and compare the results of your pruning functions to those of the python implementation.

Commit:
fixing redundancy in pruning.R

Commit:
delete unused variables weakestLink.R

parismita commented:
ok, i'll add the tests and examples asap.....

Commit:
Rename fitAndScore.R to Rpackage/R/fitAndScore.R

Commit:
Rename metric.R to Rpackage/R/metric.R

parismita committed:
created test for pruning function

parismita committed:
write test for metric.R

parismita committed:
solve error in mse formula in metric.R

parismita committed:
rename pruning.R to pruning.mmit.R and update tests for pruning.mmit.R

parismita committed:
bug fix, remove checking for float node id in pruning.mmit.R

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+                                    n_folds = 3, scorer=NULL, loss = "hinge",
+                                    pruning = TRUE){
+  ### total length
+  l <- length(target.mat[, 1])
```
Minor comment: it's probably more efficient to use l <- dim(target.mat)[1]. The dimensions are probably stored in the data structure and accessible in O(1) time, whereas your current approach performs a slice before checking the length.

aldro61 replied:
Actually, you can use the nrow function for that.

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+  fold_split_idx$test <- seq(from = ((l/n_folds)*(n_folds-1)) + 1, to = l, by = 1)
+  
+  ### middle values
+  for(n in (n_folds-1):2){
```
This code is very complicated and does not seem to work according to my tests. We need to discuss this in the meeting today.

aldro61 replied:
Splitting the data into folds can be done very simply, see https://stats.stackexchange.com/a/105839/16717

aldro61 replied:
Simply shuffle the examples before using that code, as it does not use random order.

aldro61 replied:
Also, you should be using a specific random seed in the shuffling to ensure reproducibility.

parismita replied:
done

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+  ### if pruning
+  if(pruning){
+    # Get the pruned master and cross-validation trees
+    master_data <- pruning(master_tree)
```
In general, if your functions return multiple values, you should use lists: https://stackoverflow.com/a/8936179/1172787. This way, you can use names instead of indices to access elements, which makes the code much more readable.

aldro61 replied:
alphas <- lapply(master_data, function(x) x$alpha)
trees <- lapply(master_data, function(x) x$tree)

parismita replied:
done

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+      fold_prune_trees[[i]] <- fold_data[, -1]
+    }
+    
+    ### alphas list should not contain repeating alpha
```
If this happens, it's because there is a bug in your pruning code. In the present form, the code would silently continue, which is incorrect.

parismita replied:
done

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+    # Compute the test risk for all pruned trees of each fold
+    alpha_path_score <- NULL
+    for(i in 1 : n_folds){
+      for(j in 1 : length(fold_prune_trees[[i]][,1])){
```
Minor comment: Same comment as above for the use of length. Try to use more specialized functions like nrow. It will make the code more readable.

parismita replied:
done

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+        fold_test_scores <- scorer(target.mat[fold_split_idx$test[i,],], prediction)
+        
+        ### creating a dataframe with init alpha, final alpha, score value
+        if(j < (length(fold_alphas[[i]]) - 1)){
```
Indices in R start at 1, so this condition is incorrect. I think the minus 1 should not be there.

parismita replied:
done

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+          alpha_path_score <- rbind(alpha_path_score, c(fold_alphas[[i]][j], Inf, fold_test_scores))
+        }
+        colnames(alpha_path_score) <- c("init alpha", " final alpha", "score")
+        alpha_path_score <- as.data.frame(alpha_path_score)
```
There is something that I don't understand. You are using one single alpha_path_score for all folds, but adding one row for each pruned tree of each fold. Then you are setting the colnames each time and converting it to a dataframe (each time).

In alpha_path_score, you should be able to easily access the alpha path for each fold, separately.

aldro61 replied:
Also, you can't expect to have the same number of alphas for each fold, so probably using a vector of lists or a list of lists would be more appropriate.

aldro61 replied:
Live meeting: decided to use a list of dataframes.

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+    best_tree <- NULL
+    
+    for(i in 1:length(master_pruned_trees[,1])){
+      if(i < (length(master_alphas) - 1)){
```
Check if -1 one is appropriate since indices start at 1 (same comment as above). The inf value is really for the last element only, since it's a range from [x, inf[, leading the mean to be inf.

parismita replied:
done

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+      
+      ### compute cv_score as mean of each fold scores
+      cv_score <- 0
+      for(i in 1:n_folds){
```
`i` does not even appear in this list. I really don't understand what's going on here.

aldro61 replied:
alpha_path_score should be indexed with the fold

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+        fit <- fitted_node(node, feature.mat[fold_split_idx$test[i,],])
+        n <- nodeapply(node, ids = fit, info_node)
+        ###prediction of test data
+        prediction <- matrix(unlist(n), nrow = length(n), byrow = T)[,1]
```
Why are the predictions extracted from info node? You should simply be applying the fitted mmit tree to the data to compute predictions. Doesn't partykit provide a method for computing predictions with a tree?

aldro61 replied:
Answered in live meeting. 👍

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+      ### calc train score
+      fit <- predict(master_tree, feature.mat)
+      n <- nodeapply(master_tree, ids = fit, info_node)
+      prediction <- matrix(unlist(n), nrow = length(n), byrow = T)[,1]
```
Why are you extracting the predictions from the info node?

aldro61 replied:
Answered 👍

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+    ### For each fold, build a decision tree
+    fold_test_scores <- NULL
+    for(i in 1:n_folds){
+      fit <- predict(fold_tree[[i]], feature.mat[fold_split_idx$test[i,],])
```
Same comment as above.

aldro61 commented on Rpackage/tests/testthat/test-metric.R:
```diff
@@ -0,0 +1,36 @@
+library(testthat)
```
Good! 👍

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+      prediction <- matrix(unlist(n), nrow = length(n), byrow = T)[,1]
+      train_score <- scorer(target.mat, prediction)
+      
+      ### calc cost of all leaves
```
Could do that after finding the best tree so covers pruning=True and false

parismita committed:
use nrow and ncol where required

parismita committed:
pruning -> mmit.pruning rename

parismita committed:
converting return value of mmit.pruning to a list

parismita committed:
use lapply to get alphas and trees in fitAndScore

Commit:
shuffle and segment the data in fit and score

parismita committed:
define folds (bug solve) fit and score, shufling the data

Commit:
Update test_pruning.R

parismita committed:
test_pruning, change the output of mmit.pruning to list

parismita committed:
minor bugs in fit_and_score

parismita committed:
create mmit.predict for prediction values of test data

parismita committed:
create mmit.cv for crossvalidation

aldro61 commented on Rpackage/R/weakestLink.R:
```diff
@@ -0,0 +1,88 @@
+      
+      ### current alpha value
+      current_gt <- (node_cost - C_Tt)/(length(ter_id) - 1)
+      
```
Debug: add a runtime check for negative values here. Should not happen.

aldro61 replied:
Is this solved?

aldro61 commented on Rpackage/R/mmit.cv.R:
```diff
@@ -0,0 +1,47 @@
+  parameters <- NULL
+  row <- nrow(param_grid)
+  col <- ncol(param_grid)
+  parameters$maxdepth <- rep(param_grid[,1], each = row**(col-1))
```
```expand.grid(max_depth = max_depth, margin = margin, min_sample = min_sample, loss = loss)```

parismita replied:
done

aldro61 commented on Rpackage/R/mmit.cv.R:
```diff
@@ -0,0 +1,47 @@
+      ### see all permutation combination of param grid values
+      cv_result[[i]] <- fit_and_score(target.mat = target.mat, feature.mat = feature.mat, 
+                                       parameters = parameters[i,], feature_names = feature_names, 
+                                       n_folds = n_folds, scorer = scorer, loss = loss,
```
Loss should be in the parameters (also update fit and score)

parismita replied:
done

aldro61 commented on Rpackage/R/mmit.cv.R:
```diff
@@ -0,0 +1,47 @@
+mmit.cv <- structure(function(target.mat, feature.mat, 
+                              param_grid, feature_names = NULL, 
+                              n_folds = 3, scorer = NULL, loss = "hinge",
```
Scorer is a function that is always increasing (e.g., zero one loss or -MSE).

aldro61 replied:
We should add attributes to our scoring functions so that we know if we should take the min or max. Also mention in the doc, that we expect the "direction" attribute for any user defined scoring function.

```
mse <- structure(mse <- function(y_true, y_pred){
+     error <- 0.0
+     for(i in 1 : length(y_pred)){
+         if(y_true[i,1] > y_pred[i]){
+             error <- error + (y_true[i,1] - y_pred[i])**2
+         }
+         else if(y_true[i,2] <= y_pred[i]){
+             error <- error + (y_true[i,2] - y_pred[i])**2
+         }
+     }
+     return((error / length(y_pred)))
+ }, direction=min)
```

aldro61 replied:
We also decided to add "worst" attribute which will be used in fit and score.

parismita replied:
done

aldro61 commented on Rpackage/R/mmit.cv.R:
```diff
@@ -0,0 +1,47 @@
+                                       n_folds = n_folds, scorer = scorer, loss = loss,
+                                       pruning = pruning)
+
+      if(best_result$best_score >= cv_result[[i]]$best_score){
```
```
if(attr(scorer, "direction")(best_result$best_score, cv_result[[i]]$best_score) == cv_result[[i]]$best_score){
    
}
```

parismita replied:
done

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+        fold_test_scores <- scorer(target.mat[fold_split_idx$test[i,],], prediction)
+        
+        ### creating a dataframe with init alpha, final alpha, score value
+        if(j < (length(fold_alphas[[i]]))){
```
remove extra ()

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+        ###error calc
+        fold_test_scores <- scorer(target.mat[fold_split_idx$test[i,],], prediction)
+        
+        ### creating a dataframe with init alpha, final alpha, score value
```
remove extra ()

aldro61 replied:
Also fix the indentation

parismita replied:
done

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+        ### creating a dataframe with init alpha, final alpha, score value
+        if(j < (length(fold_alphas[[i]]))){
+          if(length(alpha_path_score)<i){
+            alpha_path_score[[i]] <- c(fold_alphas[[i]][j], fold_alphas[[i]][j + 1], fold_test_scores)
```
Don't copy `c(fold_alphas[[i]][j], fold_alphas[[i]][j + 1], fold_test_scores)` twice. Same comment for the inf case

parismita replied:
done

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+    best_score <- Inf
+    best_tree <- NULL
+    
+    for(i in 1 : length(master_alphas)){
```
fix ()

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+    best_tree <- NULL
+    
+    for(i in 1 : length(master_alphas)){
+      if(i < (length(master_alphas))){
```
() fix

parismita replied:
dome

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+    alpha_train_scores <- NULL
+    alpha_train_objective_values <- NULL
+    best_alpha <- -1
+    best_score <- Inf
```
Use worst attr here.

parismita replied:
done

aldro61 commented on Rpackage/R/mmit.cv.R:
```diff
@@ -0,0 +1,47 @@
+  ### for param grid run loop to check best value
+  cv_result <- NULL
+  best_result <- NULL
+  best_result$best_score <- -Inf
```
use "worst" attribute of scorer.

parismita replied:
done

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+      cv_score <- 0
+      for(i in 1 : n_folds){
+        for(j in 1 : nrow(alpha_path_score[[i]])){
+          if((geo_mean_alpha_k < alpha_path_score[[i]][j, 2]) && (geo_mean_alpha_k >= alpha_path_score[[i]][j, 1])){
```
Use name instead of index

parismita replied:
done

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+      alpha_train_scores <- c(alpha_train_scores, train_score)
+      alpha_train_objective_values <- c(alpha_train_objective_values, train_objective)
+      
+      if(cv_score < best_score){
```
Make sure to use the scorer attributes here.

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+    alpha_train_scores <- NULL
+    alpha_train_objective_values <- NULL
+    best_alpha <- -1
+    best_score <- Inf
```
Use scorer "worst" attribute to initialize.

parismita replied:
done

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,191 @@
+  best_params$alpha <- best_alpha
+  
+  ### Generate a big dictionnary of all HP combinations considered (including alpha) and their CV scores
+  cv_results <- cbind(parameters$maxdepth, parameters$margin, parameters$min_sample, 
```
Add the loss hyperparameter here.

parismita replied:
done

aldro61 commented on Rpackage/R/mmit.predict.R:
```diff
@@ -0,0 +1,8 @@
+    
+  fit <- predict(tree, newdata, perm)
+  n <- nodeapply(tree, ids = fit, info_node)
+  prediction <- matrix(unlist(n), nrow = length(n), byrow = T)[, 1]
```
If possible use name of column

parismita replied:
done

parismita committed:
added loss as param, fixed intendations and minor fixes in fit and score

parismita committed:
added attributes in metric

parismita committed:
added loss as param, in mmit.cv

parismita committed:
remove loss from line 18, mmit.cv

parismita committed:
convert index to name in mmit.predict

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,195 @@
+      alphas <- c(alphas, geo_mean_alpha_k)
+      alpha_cv_scores <- c(alpha_cv_scores, cv_score)
+      
+      if(cv_score < best_score){
```
scorer direction should be used here.

parismita committed:
use direction attribute for comparision, fitAndScore.R

parismita committed:
include assetthat, fix bug in mmit.cv

parismita committed:
assert for positive current_gt, wealestLink.R

parismita committed:
add library doParallel and parallel to namespace and description

parismita committed:
add leftcost and rightcost to bestsplit

parismita committed:
fix growtree for new bestsplit

parismita committed:
add assertion in initial.pruning

parismita committed:
parallalize the for loop mmit.cv

parismita committed:
add foreach in namespace

parismita committed:
fix bug in test_pruning

parismita committed:
maxdepth to max_depth, remove feature_names, solve the parallalization bug.

parismita committed:
documentation

parismita committed:
change maxdepth  to max_depth in test_tree

parismita committed:
fix bug fit_and_score

parismita committed:
add assertion and default value for parameters mmit.cv

parismita committed:
fix bug, no of fold alpha equal to no of pruned trees, fit and score

parismita committed:
add cv_results, mmit.cv

parismita committed:
train score bug fix, fitAndScore

aldro61 commented on Rpackage/R/mmit.cv.R:
```diff
@@ -0,0 +1,104 @@
+  if(is.null(param_grid$loss)) param_grid$loss <- "hinge"
+  
+  ### check for unwanted parameters
+  assert_that(length(param_grid) <= 4, msg = "unexpected parameters as arguement")
```
typo: argument

parismita replied:
done

aldro61 commented on Rpackage/R/mmit.predict.R:
```diff
@@ -0,0 +1,39 @@
+#' pred <- mmit.predict(tree)
+#' 
+#' @export
+mmit.predict <- structure(function(tree, newdata = NULL, perm = NULL){
```
We don't support surrogate splits in MMIT. At least, I don't think that you implemented that?

parismita replied:
yeah...removed (i didnt know what surrogate split meant)

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,204 @@
+                                    parameters, n_folds = 3, scorer = NULL, 
+                                    pruning = TRUE){
+  ### total length
+  l <- nrow(target.mat)
```
Unused variable

parismita replied:
removed

aldro61 commented on Rpackage/R/mmit.cv.R:
```diff
@@ -0,0 +1,104 @@
+#' The Cross Validation of Max Margin Interval Tree
```
Not a big deal, but I'm curious: why is the loss type 0 or 1 in the cv_results instead of "hinge" or "squared"?

parismita replied:
it happens after the foreach statement, so I guess thats the reason, fit_and_score gives strings only, i dont know why it changed to labels here. maybe i'll have to manually change it back

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,204 @@
+  if(pruning){
+    # Get the pruned master and cross-validation trees
+    master_data <- mmit.pruning(master_tree)
+    master_alphas <- rev(unlist(lapply(master_data, function(x) x$alpha)))
```
I don't understand why you reverse the alphas and the trees. Can you explain please?

aldro61 replied:
I fact, I think that reverse sorting the alphas does not make sense and breaks the pruning code.

aldro61 replied:
See comment at line 83.

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,204 @@
+          alpha_path_score[[i]] <- rbind(data, alpha_path_score[[i]])
+        }
+        else{
+          data <- c(fold_alphas[[i]][j], Inf, fold_test_scores)
```
Your alphas are sorted in decreasing order, so something like: a1, a2, ..., 0, where 0 corresponds to no pruning at all. It does not make sense that infinity is the alpha that follows 0. 

As a reference, in the python code, the alphas and trees are not reverse sorted (for master and the folds): https://github.com/aldro61/mmit/blob/master/mmit/model_selection.py#L62

aldro61 replied:
If you iterate over the alphas in increasing order, then it makes sense that infinity is after a_n.

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,204 @@
+        geo_mean_alpha_k <- sqrt(master_alphas[[i]] * master_alphas[[i + 1]])
+      }
+      else{
+        geo_mean_alpha_k <- Inf
```
Again, this doesn't make sense if the alphas are reverse sorted.

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,204 @@
+      for(j in 1 : n_folds){
+        for(k in 1 : nrow(alpha_path_score[[j]])){
+          if((geo_mean_alpha_k < alpha_path_score[[j]][k, ]$` final alpha`) && (geo_mean_alpha_k >= alpha_path_score[[j]][k, ]$`init alpha`)){
+            cv_score <- cv_score + alpha_path_score[[j]][k, ]$score
```
In the CV results you have some 0 values in cv_score, which is highly surprising. These 0 values are always associated with alpha = inf. I suspect that this is a bug due to the reverse sorting of the alphas and trees. Should verify if this still occurs after correcting the bug.

aldro61 replied:
So as you said, alphas are in correct order, but I still find it odd to have some cv scores equal to 0. I'll debug to see if it happens that you don't get into the loop

aldro61 replied:
Can you please verify that the score is indeed updated for each fold? This can be done by setting a variable to False at the beginning of the loop of a fold (e.g., line 122) and setting that variable to True once the condition is satisfied. Then at the end of the fold loop, you can assert that the condition was satisfied at least once for the fold.

I suspect that sometimes the code doesnt go into the condition and thus the score is left at 0.

parismita replied:
yeah...possibly it doeant go into the loop for alpha = Inf, i'll check

parismita replied:
found the bug, it was not going inside the loop due to this statement
`if((geo_mean_alpha_k < alpha_path_score[[j]][k, ]$final alpha` 
When geo_mean_alpha_k  = Inf this gives false even for final alpha = Inf, thus it doesnt go inside the loop.

parismita replied:
But what values does the between dict consist of? is it initial alpha to final alpha excluding final alpha or including, because in next (j+1th) iteration the final alpha of j is initial alpha of j+1th loop

aldro61 replied:
So in the Python implementation, the between dict is used to find the pruned tree that corresponds to the range [alpha_k, alpha_(k+1)[ (notice exclusion of upper limit). So if the geomean is in that range, the corresponding tree is returned. You can see that the between dict has special conditions for infinite values: https://github.com/aldro61/mmit/blob/master/mmit/utils.py#L39. Basically, you should reproduce this condition in your code. In the new code (after your bug fix), you have the condition `geo_mean_alpha_k <= alpha_path_score[[j]][k, ]$` final alpha`, which I think is incorrect, because of the `=` and the fact that the upper limit should be excluded.

parismita replied:
rectified

aldro61 commented on Rpackage/R/mmit.pruning.R:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    for(n in nrow(wlink):1){
+      tree <- nodeprune(tree, ids = wlink[n, 2])
```
Why don't you simply call nodeprune once with all ids to prune? The function can handle a vector of ids. You would avoid all potential bugs due to IDs changing after pruning.

parismita replied:
ok

aldro61 commented on Rpackage/R/mmit.cv.R:
```diff
@@ -0,0 +1,104 @@
+  if(is.null(param_grid$loss)) param_grid$loss <- "hinge"
+  
+  ### check for unwanted parameters
+  assert_that(length(param_grid) <= 4, msg = "unexpected parameters as arguement")
```
This condition does not work. I discovered this by mistake. 

There is a very weird feature in R that allows to match elements in list with partial names. For instance, if the list contains an entry with name "alex", then you can get it with mylist$al and also mylist$ale, and also mylist$alex.

So basically, when I created a param_grid with min_samples instead of min_sample, it was passing validation properly, but failing later in the code.

The solution is to replace mylist$alex with mylist[[alex]] over here.

parismita replied:
done

aldro61 commented:
Do you think it would be better to use namespacing when calling functions inside mmit? For instance, in mmit.cv, you call fit_and_score. What if another user has defined fit_and_score? It might be more robust to use mmit::fit_and_score. However, I'm not familiar with the best practices. This is just my intuition.

parismita committed:
remove unused variable fit_and_score, update doc mmit.predict, remove loop of mmit.pruning

parismita committed:
fix bug, line 71 weakest link, left_weakest_link to be returned

parismita committed:
change from list$arguement to list[["arguement"]] for if condition, mmit.cv

parismita commented:
Yeah, this way there wont be any namespace conflicts

parismita commented:
Nope, its automatically taken care of by NAMESPACE file, I just checked defining some global function with same name, but the functions used by mmit already use mmit as namespace.

parismita committed:
bug fix, alpha = Inf lead to cv_score = 0, fit_and_score

parismita committed:
fix bug: splitting cost is more than current node cost, growtree

parismita committed:
extra condition to check geo_mean_alpha = Inf, fit_and_score

parismita committed:
Changing dataset in depth test to get larger tree, test-tree.R

parismita committed:
Changed namespace and description to import penaltyLearning
package survival not used in mmit, package penaltyLearning needed for dataset

aldro61 committed:
Update fitAndScore.R

aldro61 commented on Rpackage/R/fitAndScore.R:
```diff
@@ -0,0 +1,206 @@
+      fold_data <- mmit.pruning(fold_tree[[i]])
+      
+      ### as mmit.pruning gives the alphas in descending order we need to reverse the list.
+      fold_alphas[[i]] <- rev(unlist(lapply(fold_data, function(x) x$alpha)))
```
Getting more elements supplied than there are to replace error. Need to investigate.

parismita replied:
done

aldro61 commented:
Actions items:

* Create a benchmark of a few datasets (at least 10) using git.io/mmit-data and compare the MSE of the python and R implementations on a testing set (50% train - 50% test)
* Fix the error message in fit_and_score (fold alphas)
* Reevaluate objectives once the benchmark results are done.
* Check if alpha=inf is often the best for other datasets (would indicate that something is wrong)

parismita committed:
bug fix: initialise list

Commit:
Create pred_score.md

Commit:
Rename Rpackage/benchmark/pred_score.md to benchmark/pred_score.md

parismita committed:
code for benchmark

parismita committed:
code fo benchmark (50% split)

Commit:
Update pred_score.md

Commit:
Update pred_score.md

Commit:
Update pred_score.md

aldro61 commented on benchmark/pred_score.md:
```diff
@@ -0,0 +1,34 @@
+| Dataset       | Python             | R        | size |  comment
+| ------------- |:-----------------: | ---------:|---: |-----: |
+| auto93        | 0.478885004573     | 0.007852035 | 82| initially same |
```
What does initially same mean?

parismita replied:
means the tree was same for depth = 2 or 3, and later it was different

aldro61 commented on benchmark/pred_score.md:
```diff
@@ -0,0 +1,34 @@
+| Dataset       | Python             | R        | size |  comment
```
You should really focus on the largest datasets. The size of each dataset is reported in the supp mat of the mmit paper: https://papers.nips.cc/paper/7080-maximum-margin-interval-trees-supplemental.zip.

parismita replied:
ok

aldro61 commented on benchmark/pred_score.md:
```diff
@@ -0,0 +1,34 @@
+
+
+| Dataset       | Python time            | R time 
+| ------------- |:-----------------: | ---------:|
```
Ouch, thats a big difference in running time. Lets not focus on fixing the speed issues for now. We can do that later.

parismita replied:
yeah....when we use for loops in R it gives larger time than for loops in python, so replacing them might reduce the time.

aldro61 commented on benchmark/score.R:
```diff
@@ -0,0 +1,25 @@
+  test.tar <- data.matrix(tail(target.mat, nrow(target.mat)/2))
+  
+  
+  tree <- mmit(train.tar, train.feat, max_depth = 4, margin = 1.0)
```
I know that the default parameter values match between the python and r implementation, but still specify all parameters to be sure. Could have been a cause of the difference.

parismita replied:
ok

aldro61 commented on benchmark/mmitt.py:
```diff
@@ -0,0 +1,61 @@
+# Load CSV (using python)
```
Typo in the name mmit(t)

parismita replied:
no , actually the name cant match the package so i made the typo knowingly :P

parismita replied:
i'll change it to something else

aldro61 commented on benchmark/mmitt.py:
```diff
@@ -0,0 +1,61 @@
+			'/home/parismita/mmit_data/sleep/features.csv', '/home/parismita/mmit_data/sleep/targets.csv']
+
+
+for i in xrange(20):
```
You could have simply used this class to load the data: https://github.com/tdhock/mmit-paper/blob/master/mmit.predictions.py#L23

aldro61 replied:
You can make the code much cleaner:
```
from os import *
from os.path import *

def find_datasets(path):
    for d in listdir(path):
        if exists(join(path, d, "features.csv")) and \
                exists(join(path, d, "targets.csv")) and \
                exists(join(path, d, "folds.csv")):
            yield Dataset(abspath(join(path, d)))

for dataset in find_datasets("/home/parismita/mmit_data"):
    X = d.X
    y = d.y
    # Then split X and y in half and train the algorithm here
```

parismita replied:
ok

aldro61 commented on benchmark/mmitt.py:
```diff
@@ -0,0 +1,61 @@
+			'/home/parismita/mmit_data/sleep/features.csv', '/home/parismita/mmit_data/sleep/targets.csv']
+
+
+for i in xrange(20):
```
You could have simply used this class to load the data: https://github.com/tdhock/mmit-paper/blob/master/mmit.predictions.py#L23

aldro61 replied:
You could have simply used this class to load the data: https://github.com/tdhock/mmit-paper/blob/master/mmit.predictions.py#L23. 

I believe that the code you wrote works, but it is very complicated (using many indexes and loops).

aldro61 replied:
You could have simply used this class to load the data: https://github.com/tdhock/mmit-paper/blob/master/mmit.predictions.py#L23. 

I believe that the code you wrote works, but it is very complicated (using many indexes and loops).

aldro61 commented on benchmark/mmitt.py:
```diff
@@ -0,0 +1,61 @@
+	testx = x[len(x)/2:,]
+	testy = y[len(y)/2:,]
+
+	estimator = MaxMarginIntervalTree(margin=1.0, max_depth=4)
```
Same comment for the default parameters, set everything

parismita replied:
ok

aldro61 commented on benchmark/mmitt.py:
```diff
@@ -0,0 +1,61 @@
+	testx = x[len(x)/2:,]
+	testy = y[len(y)/2:,]
+
+	estimator = MaxMarginIntervalTree(margin=1.0, max_depth=4)
```
Same comment for the default parameters, set everything

aldro61 commented:
Ok, I reviewed the benchmark code. It looks fine (see my comments). I noticed that you don't shuffle the data, which could be a problem if the examples are ordered in a particular way, but I think it is ok here, since we are just trying to compare both implementations on the same data.

Please do the comparison with the largest datasets only. The models and accuracies obtained for small datasets are likely to be very unstable.

parismita commented:
I didnt shuffle them, so that same data goes into training for both implementations so that trees have higher probability to match

Commit:
Update pred_score.md

Commit:
Update mmitt.py

Commit:
Update score.R

Commit:
Add files via upload

aldro61 commented on Rpackage/R/growtree.R:
```diff
@@ -1,38 +1,49 @@
-
+  
+  ### tree cant grows if splitting cost is more than current node cost
+  assert_that(node_info_print$cost >= (sp$leftcost + sp$rightcost))
```
Equal should not be in this condition since if its equal then you're better off not splitting at all.
